lib/generic.js

```diff
@@ -0,0 +1,153 @@
+/*!
+ * body-parser
+ * Copyright(c) 2014 Jonathan Ong
+ * Copyright(c) 2014-2015 Douglas Christopher Wilson
+ * MIT Licensed
+ */
+
+'use strict'
+
+/**
+ * Module dependencies.
+ * @private
+ */
+
+const contentType = require('content-type')
+const createError = require('http-errors')
+const debug = require('debug')('body-parser:generic')
+const { isFinished } = require('on-finished')
+const read = require('./read')
+const typeis = require('type-is')
+const utils = require('./utils')
+
+/**
+ * Module exports.
+ * @public
+ */
+
+module.exports = generic
+
+/**
+ * Create a middleware to parse request bodies.
+ *
+ * @param {Object} [options]
+ * @param {Function} [options.parse] Function to parse body (required). This function:
+ *   - Receives (buffer, charset) as arguments
+ *   - Must be synchronous (cannot be async or return a Promise)
+ *   - Will be called for requests with empty bodies (zero-length buffer)
+ *   - Will NOT be called for requests with no body at all (e.g., typical GET requests)
+ *   - Return value becomes req.body
+ * @param {String|Function} [options.type] Request content-type to match (required)
+ * @param {String|Number} [options.limit] Maximum request body size
+ * @param {Boolean} [options.inflate] Enable handling compressed bodies
+ * @param {Function} [options.verify] Verify body content
+ * @param {String} [options.defaultCharset] Default charset when not specified
+ * @param {String} [options.charset] Expected charset (will respond with 415 if not matched)
+ * @return {Function} middleware
+ * @public
+ */
+
+function generic (options) {
+  // === STEP 0: VALIDATE OPTIONS ===
```

Comment:
Breaking this function down into steps was very helpful for me when wrapping my head around what was a universal concern vs parser specfic. leaving them here hoping they help others